### PR TITLE
Wasm POC

### PR DESCRIPTION
**Pull request by sarawut11**
WASM implementation for powersoftau new - creates a new challenge for the ceremony

### CODE REVIEW

File: snark-utils/Cargo.toml [hunk 1]
```diff
@@ -38,5 +38,6 @@ criterion = { version = "0.3.1" }
 rusty-hook = { version = "0.11.2" }
 
 [features]
-default = []
-parallel = ["rayon", "zexe_algebra/parallel", "zexe_fft/parallel"]
+default = ["rust-crypto"]
```
Comment (howardwu) [inline snark-utils/Cargo.toml:41]: Is there a reason "parallel" is not enabled by default?

File: snark-utils/src/helpers.rs [hunk 2]
```diff
@@ -127,6 +128,7 @@ pub fn user_system_randomness() -> Vec<u8> {
     arr.to_vec()
 }
 
+#[cfg(feature = "default")]
```
Comment (howardwu) [inline snark-utils/src/helpers.rs:131]: Is this only supposed to be enabled on "default" and not on "parallel" as well?

File: wasm-wrapper/tests/web.rs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+#![cfg(target_arch = "wasm32")]
+
+extern crate wasm_bindgen_test;
+use wasm_bindgen_test::*;
+
+wasm_bindgen_test_configure!(run_in_browser);
+
+#[wasm_bindgen_test]
+fn pass() {
```
Comment (howardwu) [inline wasm-wrapper/tests/web.rs:11]: You can remove this test

File: powersoftau/Cargo.toml [hunk 4]
```diff
@@ -13,19 +13,22 @@ name = "accumulator"
...
+memmap = { version = "0.7.0", optional = true }
 itertools = { version = "0.8.0" }
 rand = { version = "0.7" }
 rayon = { version = "1.3.0" }
 tracing = { version = "0.1.17" }
 tracing-subscriber = { version = "0.2.3" }
 
+[features]
+default = ["memmap"]
+wasm = []
```
Comment (howardwu) [inline powersoftau/Cargo.toml:30]:
```suggestion
wasm = ["snark-utils/wasm"]
```

File: wasm-wrapper/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        };
+        return Ok(JsValue::from_serde(&res.ok().unwrap()).unwrap());
+    }
+}
+
+pub fn get_parameters<E: Engine>(batch_size: usize, power: usize) -> CeremonyParams<E> {
+    CeremonyParams::<E>::new(power, batch_size)
+}
+
+pub fn new_challenge<T: Engine + Sync>(parameters: &CeremonyParams<T>) -> Result<ChallengeInfo, String> {
```
Comment (kobigurk) [inline wasm-wrapper/src/lib.rs:46]: We don't need `new_challenge` (nor `verify`) in WASM.
We only need `contribute` for both phases.

File: powersoftau/src/cli_common/mod.rs [hunk 6]
```diff
@@ -1,10 +1,16 @@
+#[cfg(feature = "default")]
```
Comment (kobigurk) [inline powersoftau/src/cli_common/mod.rs:1]: How about using `cfg_if` here?

File: powersoftau/Cargo.toml [hunk 7]
```diff
@@ -13,19 +13,22 @@ name = "accumulator"
...
 [dependencies]
-snark-utils = { path = "../snark-utils", features = ["parallel"] }
-
+snark-utils = { path = "../snark-utils" }
 zexe_algebra = { git = "https://github.com/scipr-lab/zexe", package = "algebra", version = "0.1.0", features = ["parallel", "bw6_761", "derive"] }
 
 gumdrop = { version = "0.7.0" }
 hex-literal = { version = "0.1.4" }
-memmap = { version = "0.7.0" }
+memmap = { version = "0.7.0", optional = true }
```
Comment (kobigurk) [inline powersoftau/Cargo.toml:21]: Does the code have a way to remove the `memmap` now?

**Comment by kobigurk** [comment]
Is this PR still relevant given #26?

**Comment by sarawut11** [comment]
yes, let me close this

**Pull request closed without merging**